### ISSUE TITLE
Unable to connect to Ubuntu server from Vscode on macbook using remote-ssh

### ISSUE THREAD

**Issue by ramanNarasimhan77** [microsoft/vscode-remote-release#4176]
## VSCode version: 
1.51.1
e5a624b788d92b8d34d1392e4c4d9789406efe8f
x64

## Issue

I am trying to connect to a ubuntu server from vscode running in a macbook.  SSH connection from macbook to ubuntu server can be made using an identity file, and this has been verified from terminal.

```
Host redacted
    Hostname redactedUbuntuHost
    User redactedUser
    IdentityFile ~/.ssh/id_rsa
```

However when I try to use the Remote-ssh extension and make a connection to the server; 
VSCode shows the message "Failed to connect to the remote extension host server (Error: Time limit reached)"
and then shows the error message "Could not fetch the remote environment"

I have increased the connection timeout as suggested in the troubleshooting steps, but still face this issue.

When I start the session in vscode, I can see the following logs,

```
717b5473e20d: running
Acquiring lock on /home/usernameRedacted/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f/vscode-remote-lock.usernameRedacted.e5a624b788d92b8d34d1392e4c4d9789406efe8f
Installing to /home/usernameRedacted/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f...
717b5473e20d%%1%%
Downloading with wget
Download complete
717b5473e20d%%2%%
tar --version:
tar (GNU tar) 1.30
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
Checking /home/usernameRedacted/.vscode-server/.e5a624b788d92b8d34d1392e4c4d9789406efe8f.log and /home/usernameRedacted/.vscode-server/.e5a624b788d92b8d34d1392e4c4d9789406efe8f.pid for a running server
Starting server with command... /home/usernameRedacted/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f/server.sh --host=127.0.0.1 --enable-remote-auto-shutdown  --port=0 &> "/home/usernameRedacted/.vscode-server/.e5a624b788d92b8d34d1392e4c4d9789406efe8f.log" < /dev/null
printenv:
    SHELL=/bin/bash
    PWD=/home/usernameRedacted
    LOGNAME=usernameRedacted
    XDG_SESSION_TYPE=tty
    MOTD_SHOWN=pam
    HOME=/home/usernameRedacted
    LANG=en_US.UTF-8
    VSCODE_AGENT_FOLDER=/home/usernameRedacted/.vscode-server
    SSH_CONNECTION=ip1Redacted 54773 ip2Redacted 22
    XDG_SESSION_CLASS=user
    USER=usernameRedacted
    SHLVL=0
    XDG_SESSION_ID=97
    XDG_RUNTIME_DIR=/run/user/1003
    SSH_CLIENT=ip1Redacted 54773 22
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
    DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1003/bus
    OLDPWD=/home/usernameRedacted/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f
    _=/usr/bin/printenv
Spawned remote server: 272459
Waiting for server log...
 
*
* Reminder: You may only use this software with Visual Studio family products,
* as described in the license (https://go.microsoft.com/fwlink/?linkid=2077057)
*
 
717b5473e20d: start
webUiAccessToken====
listeningOn==39185==
osReleaseId==ubuntu==
arch==x86_64==
tmpDir==/run/user/1003==
platform==linux==
unpackResult==success==
didLocalDownload==0==
downloadTime==402==
installTime==706==
extInstallTime====
serverStartTime==208==
717b5473e20d: end
```

I can see the following process running on the remote host

```
redactedUser@redactedHostname:~$ ps -ef | grep vscode
redactedUser   272459  272399  0 09:50 ?        00:00:00 sh /home/redactedUser/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f/server.sh --host=127.0.0.1 --enable-remote-auto-shutdown --port=0
redactedUser   272466  272459  0 09:50 ?        00:00:00 /home/redactedUser/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f/node /home/redactedUser/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f/out/vs/server/main.js --host=127.0.0.1 --enable-remote-auto-shutdown --port=0
```

Also, here is the server log

```
redactedUser@redactedHostname:~$ cat ~/.vscode-server/.e5a624b788d92b8d34d1392e4c4d9789406efe8f.log


*
* Visual Studio Code Server
*
* Reminder: You may only use this software with Visual Studio family products,
* as described in the license https://aka.ms/vscode-remote/license
*


IP Address: redacted
Extension host agent listening on 39185


[09:50:19] Extension host agent started.
```

Is this caused by incorrect settings or a bug?

**Comment by roblourens** [comment]
What's your remote OS version? x64/arm?

**Comment by ramanNarasimhan77** [comment]
> What's your remote OS version? x64/arm?

## OS version
```
redactedUser@redactedHostname::~$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

## Arch

```
redactedUser@redactedHostname:~$ uname -a
Linux redactedHostname 5.4.0-48-generic #52-Ubuntu SMP Thu Sep 10 10:58:49 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by roblourens** [comment]
Seems that everything was set up successfully, but vscode can't open a connection to your server.

Can you also share the log from the Remote-SSH output channel?

Possibilities
- Port forwarding is not enabled for your SSH server
- We couldn't forward the port successfully
- The ssh port forwarding by some network configuration on one of your systems

**Comment by ramanNarasimhan77** [comment]
@roblourens 
Sorry for my delayed response, I got caught up in some other work.

Here's the log from Remote-SSH output channel.


```
[12:14:24.842] Log Level: 2
[12:14:24.843] remote-ssh@0.62.0
[12:14:24.843] darwin x64
[12:14:24.844] SSH Resolver called for "ssh-remote+redactedRemoteServer", attempt 1
[12:14:24.844] "remote.SSH.useLocalServer": false
[12:14:24.845] "remote.SSH.showLoginTerminal": true
[12:14:24.845] "remote.SSH.remotePlatform": {"dev-master1.mee.ipos.gic.ericsson.se":"linux","redactedRemoteServer":"linux","redactedRemoteServer-pdu":"linux"}
[12:14:24.845] "remote.SSH.sshPath": undefined
[12:14:24.845] "remote.SSH.sshConfigurationFile": ~/.ssh/config
[12:14:24.845] "remote.SSH.useFlock": true
[12:14:24.845] "remote.SSH.lockfilesInTmp": true
[12:14:24.845] "remote.SSH.localServerDownload": auto
[12:14:24.845] "remote.SSH.remoteServerListenOnSocket": true
[12:14:24.846] "remote.SSH.showLoginTerminal": true
[12:14:24.846] "remote.SSH.maxReconnectionAttempts": 9007199254740991
[12:14:24.846] SSH Resolver called for host: redactedRemoteServer
[12:14:24.846] Setting up SSH remote "redactedRemoteServer"
[12:14:24.857] Using commit id "ea3859d4ba2f3e577a159bc91e3074c5d85c0523" and quality "stable" for server
[12:14:24.859] Install and start server if needed
[12:14:24.862] Checking ssh with "ssh -V"
[12:14:24.871] > OpenSSH_8.1p1, LibreSSL 2.7.3

[12:14:24.874] Using SSH config file "/Users/redactedUser/.ssh/config"
[12:14:24.874] Running script with connection command: ssh -T -D 52383 -o ConnectTimeout=60 -F '/Users/redactedUser/.ssh/config' redactedRemoteServer bash
[12:14:26.695] > 511f9fd9ed88: running
> Acquiring lock on /run/user/1003/vscode-remote-lock.redactedRemoteUser.ea3859d4ba2f3e577a159bc91e3074c5d85c0523
> Found existing installation at /home/redactedRemoteUser/.vscode-server/bin/ea3859d4ba2f3e577a159bc91e3074c5d85c0523...
> Checking /home/redactedRemoteUser/.vscode-server/.ea3859d4ba2f3e577a159bc91e3074c5d85c0523.log and /home/redactedRemoteUser/.vscode-server/.ea3859d4ba2f3e577a159bc91e3074c5d85c0523.pid for a running server
> Looking for server with pid: 1140128
> Found running server...
>  
> *
> * Reminder: You may only use this software with Visual Studio family products,
> * as described in the license (https://go.microsoft.com/fwlink/?linkid=2077057)
> *
>  
> Checking server status on socket /run/user/1003/vscode-ssh-remote-server-sock-1608446116 with curl
> 511f9fd9ed88: start
> sshAuthSock====
> webUiAccessToken====
> listeningOn==/run/user/1003/vscode-ssh-remote-server-sock-1608446116==
> osReleaseId==ubuntu==
> arch==x86_64==
> tmpDir==/run/user/1003==
> platform==linux==
> unpackResult====
> didLocalDownload==0==
> downloadTime====
> installTime====
> extInstallTime====
> serverStartTime====
> 511f9fd9ed88: end
[12:14:26.695] Got some output, clearing connection timeout
[12:14:26.695] Received install output: 
sshAuthSock====
webUiAccessToken====
listeningOn==/run/user/1003/vscode-ssh-remote-server-sock-1608446116==
osReleaseId==ubuntu==
arch==x86_64==
tmpDir==/run/user/1003==
platform==linux==
unpackResult====
didLocalDownload==0==
downloadTime====
installTime====
extInstallTime====
serverStartTime====

[12:14:26.696] Remote server is listening on /run/user/1003/vscode-ssh-remote-server-sock-1608446116
[12:14:26.696] Parsed server configuration: {"remoteListeningOn":{"socketPath":"/run/user/1003/vscode-ssh-remote-server-sock-1608446116"},"osReleaseId":"ubuntu","arch":"x86_64","webUiAccessToken":"","sshAuthSock":"","tmpDir":"/run/user/1003","platform":"linux"}
[12:14:26.698] Using SSH config file "/Users/redactedUser/.ssh/config"
[12:14:26.698] Running script with connection command: ssh -T -L 127.0.0.1:52386:/run/user/1003/vscode-ssh-remote-server-sock-1608446116 -o ConnectTimeout=60 -F /Users/redactedUser/.ssh/config redactedRemoteServer bash
[12:14:26.698] Spawning tunnel with: ssh -T -L 127.0.0.1:52386:/run/user/1003/vscode-ssh-remote-server-sock-1608446116 -o ConnectTimeout=60 -F /Users/redactedUser/.ssh/config redactedRemoteServer bash << 'EOSSH'
echo SSH_AUTH_SOCK=$SSH_AUTH_SOCK && echo -e 'Connected to SSH Host - Please do not close this terminal' && while true; do sleep 180; echo -n ' '; done
EOSSH
[12:14:27.331] > SSH_AUTH_SOCK=
> Connected to SSH Host - Please do not close this terminal
[12:14:27.331] Spawned SSH tunnel between local port 52386 and remote target /run/user/1003/vscode-ssh-remote-server-sock-1608446116
[12:14:27.331] Waiting for ssh tunnel to be ready
[12:14:27.333] Tunneled /run/user/1003/vscode-ssh-remote-server-sock-1608446116 to local port 52386
[12:14:27.333] Resolved "ssh-remote+redactedRemoteServer" to "127.0.0.1:52386"
[12:14:27.337] ------




```

**Comment by ramanNarasimhan77** [comment]
@roblourens 
Based on your last comment, I started debugging sshd settings and found that Port forwarding was in-fact turned off.

I got it to work by using the below sshd settings. 
If there are any other recommended settings kindly do let me know.

Thank you for your suggestions for debugging this issue.

<img width="264" alt="Screenshot 2020-12-20 at 1 11 47 PM" src="https://user-images.githubusercontent.com/52691361/102708665-2a927380-42ca-11eb-8e89-a41ebe43517c.png">
